lang/test/org/partiql/lang/eval/EvaluatingCompilerDateTimeTests.kt

```diff
@@ -34,4 +63,187 @@ class EvaluatingCompilerDateTimeTests : EvaluatorTestBase() {
+            case("TIME (2) WITH TIME ZONE '01:01:12' IS TIME", "true"),
+            case("'01:01:12' IS TIME", "false")
+        )
+    }
+
+    private val RANDOM_GENERATOR = generateRandomSeed()
+
+    private fun generateRandomSeed() : Random {
+        val seed = Random(1).nextInt()
+        println("Randomly generated seed is ${seed}. Use this to reproduce failures in dev environment.")
+        return Random(seed)
+    }
+
+    data class Time(val hour: Int, val minute: Int, val second: Int, val nano: Int = 0, val precision: Int? = null, val tz_minutes: Int? = null) {
+        fun expectedTimeString(withTimeZone: Boolean): String {
+            val nanoWithPrecision = when  {
+                precision == null || precision >= 9 -> nano
+                else -> (BigDecimal(nano * 10.0.pow(-9)).setScale(precision, RoundingMode.HALF_UP) * 10.0.pow(9)).toInt()
+            }
+            val newNano = nanoWithPrecision % 10.0.pow(9).toInt()
+            val newSecond = second + (nanoWithPrecision / 10.0.pow(9).toInt())
+            return when(withTimeZone) {
+                true -> {
+                    val timezoneOffsetMinutes = tz_minutes ?: ZoneOffset.systemDefault().rules.getOffset(Instant.now()).totalSeconds / 60
+                    OffsetTime.of(hour, minute, newSecond, newNano, ZoneOffset.ofTotalSeconds(timezoneOffsetMinutes * 60))
+                        .format(DateTimeFormatter.ofPattern("HH:mm:ss.nnnnnnnnnxxx"))
+                }
+                false -> LocalTime.of(hour, minute, newSecond, newNano).format(DateTimeFormatter.ofPattern("HH:mm:ss.nnnnnnnnn"))
+            }
+        }
+
+        override fun toString(): String {
+            val hourStr = hour.toString().padStart(2, '0')
+            val minStr = minute.toString().padStart(2,'0')
+            val secStr = second.toString().padStart(2, '0')
+            val nanoStr = nano.toString().padStart(9, '0')
+            val timezoneStr = tz_minutes?.let { "" +
+                (if (it >= 0) "+" else "-") +
+                (it.absoluteValue / 60).toString().padStart(2, '0') +
+                ":" +
+                (it.absoluteValue % 60).toString().padStart(2, '0')
+            } ?: ""
+            return "$hourStr:$minStr:$secStr.$nanoStr$timezoneStr"
+        }
+    }
+
+    private fun Random.nextTime(withPrecision: Boolean = false, withTimezone: Boolean = false) : Time {
+        val hour = nextInt(24)
+        val minute = nextInt(60)
+        val second = nextInt(60)
+        val nano = nextInt(999999999)
+        val precision = if (withPrecision) {
+            nextInt(20)
+        } else {
+            null
+        }
+        val timezoneMinutes = if (withTimezone) {
+            nextInt(-1080, 1081)
+        } else {
+            null
+        }
+        return Time(hour, minute, second, nano, precision, timezoneMinutes)
+    }
+
+    private val RANDOM_TIMES = List(RANDOM_TESTS_SIZE) {
+        RANDOM_GENERATOR.nextTime(
```

Comment:
Just to verify, this is deterministic (meaning each run of the tests are repeatable), yes?

Reply:
That is correct. The tests are deterministic.

Reply:
I'd prefer to make them non-deterministic actually because better test coverage.  Reproducibility is the reason you `println` the seed that was used, above.  

Yes, non-determinism can lead to intermittent failures, but that is the point.  All the tests should always succeed even if the cases being fed are deterministic.